### DISCUSSION TITLE
How to load assets in tests

### DISCUSSION THREAD

**Discussion by atomrc** [vitest-dev/vitest#6402] [Q&A]
I'm currently struggling with trying to load `assets` in `vitest`. 

Here is the gist of it:

In the source file I have 

```js
import wasmUrl from "package/package.wasm?url"
// ....

const package = fetch(new URL(wasmURL, import.meta.url)) // [1]
```

The problem is that the line `[1]`:
- resolves to `https://localserver/@fs/path/to/file.wasm` in the browser
- resolves to `file:///@fs/path/to/file.wasm` in node (and thus in the tests)

The problem is, `node` do not support using `fetch` to load local files (see https://github.com/nodejs/node/issues/45798)

I currently have a solution where I re-write the `node` `fetch` to handle the special case of `file://`. But this seems to make `msw` not work. 

I'm wondering if there would be a clean solution to avoid manual overloading and still allow nodejs tests to run. 

Any help appreciated 🙏

**Comment by atomrc** [comment]
This thread seems to indicate that the resolution to `file://` is expected in order for web worker to work (see https://github.com/vitest-dev/vitest/issues/3935). But this then conflicts with assets that are manually `fetch`ed.

**Comment by hi-ogawa** [comment]
If you have `fetch + new URL` code packaged already, then I don't think there's any way to load it on Node other than either patching `fetch` or transform the exact syntax `fetch(new URL())`.

Recently I just happen to experiment with transform approach and it's relatively simple to do this with a few regex. Probably this plugin is not directly usable for your case, but I'll link it here if you're interested https://github.com/hi-ogawa/vite-plugins/tree/main/packages/server-asset
